python/ql/src/experimental/dataflow/internal/Attributes.qll

```diff
@@ -16,13 +17,20 @@ abstract class AttrRef extends Node {
   abstract Node getObject();
 
   /**
-   * Gets the expression control flow node that defines the attribute being accessed. This is
+   * Gets the expression node that defines the attribute being accessed, if any. This is
    * usually an identifier or literal.
    */
   abstract ExprNode getAttributeNameExpr();
 
   /** Holds if this attribute reference may access an attribute named `attrName`. */
-  predicate mayHaveAttributeName(string attrName) { none() }
+  predicate mayHaveAttributeName(string attrName) {
```

Comment:
Is there a reason to not call this `hasAttributeName`? Are we any more uncertain of the values reported here than in `getAttributeName` or `getAttributeNameExpr`?
I am asking because I think this is the surface predicate, and we want to encourage users to use this one for normal use. Perhaps we should also mention this in the comment?

Reply:
Main reason is: to mimic the [JavaScript API](https://github.com/github/codeql/blob/main/javascript/ql/src/semmle/javascript/dataflow/DataFlow.qll#L501). I think the `may` is useful in this case (at least with how the library functions currently) in that `getAttributeName` will only yield a result if the name of the attribute is fixed, but `mayHaveAttributeName` can hold for several attribut names. Thus, I would expect something like 

```python
if random.randint(0,1):
    attr = "foo"
else:
    attr = "bar"
x = getattr(object, attr)
```
to have (at least) two values for `mayHaveAttributeName`.

Actually, this has me wondering. Perhaps `getAttributeName` should be rewritten to do the same kind of local flow as `mayHaveAttributeName`, but only yield a value if the result is unique. :thinking:

Reply:
In CodeQL, I have come to expect that a predicate named `getAttributeName` may yield more than one value. Is the case where the name of the attribute is fixed important?

Reply:
Ultimately, I think this will depend on whether we end up seeing false positives because of attribute confusion. I can certainly see an argument for using `mayHaveAttributeName` in something like type tracking, where we want to propagate types as much as possible (even at the cost of a bit of imprecision), but I can also imagine a situation where conflating two attribute names leads to an erroneous flow of taint. (So in particular, the data flow library itself should probably use `getAttributeName` for precision.)

Reply:
So, I was curious, and added a test case to see if we were getting the intended behaviour for `mayHaveAttributeName`, and it seems we're not.
```python
def setattr_indirect_multiple_write():
    if random.randint(0,1):
        attr = "foo"
    else:
        attr = "bar"
    x = SomeClass() # $tracked=foo $f-:tracked=bar
    setattr(x, attr, tracked) # $tracked $tracked=foo $f-:tracked=bar
```
Note the `f-` annotations above. It seems we only consider local flow from `attr = foo` and not `attr = bar`. This is true even if I negate the conditional, so I expect it's simply _always_ picking the first branch. This feels like it might be a bug in the implementation of local flow.

Reply:
Good catch :+1:

Reply:
Hm, yes, we will have to sort that out...

Reply:
If you try the same with a conditional expression, do you get the expected behaviour?

Reply:
The following test code passes (with `f-` annotations):
```python
def setattr_indirect_multiple_write_ifexpr():
    attr = "foo" if random.randint(0,1) else "bar"
    x = SomeClass() # $tracked=foo $f-:tracked=bar
    setattr(x, attr, tracked) # $tracked $tracked=foo $f-:tracked=bar
```
So, no. Same behaviour. :frowning_face:
(Also this was after manually applying https://github.com/github/codeql/commit/0f077f5d7d483a04e5be899b28ffe521a2941f7f manually, since that commit is not present on this branch.)

I'm wondering if the problem is elsewhere, though. I'll have to debug this a bit.